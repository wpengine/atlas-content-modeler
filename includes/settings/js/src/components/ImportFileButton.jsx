import React from "react";
import { __ } from "@wordpress/i18n";
import { showError } from "../toasts";
import { DarkButton } from "../../../../shared-assets/js/components/Buttons";

export default function ImportFileButton({
	errorMessage,
	buttonTitle,
	callbackFn,
	buttonClasses,
	allowedMimeTypes,
	fileUploaderRef,
}) {
	const form = new FormData();

	/**
	 * Import file click handler
	 * @param event
	 */
	function importClickHandler(event) {
		event.preventDefault();
		fileUploaderRef.current.click();
	}

	/**
	 * Validate the JSON can be successfully parsed
	 * @param file
	 * @returns {any}
	 */
	function isValidJson(file) {
		try {
			JSON.parse(file);
		} catch (e) {
			return false;
		}
		return true;
	}

	/**
	 * Validate file before uploading
	 * @param uploadedFile
	 */
	async function validateFileUpload(uploadedFile) {
		const reader = new FileReader();
		reader.addEventListener("load", async (event) => {
			if (event.target.result && isValidJson(event.target.result)) {
				form.append("file", uploadedFile);
				if (callbackFn && form) {
					await callbackFn(event.target.result);
				} else {
					showError(
						errorMessage ||
							__(
								"There was an error during upload.",
								"atlas-content-modeler"
							)
					);
				}
			} else {
				showError(
					errorMessage ||
						__(
							"The file could not be read or is invalid.",
							"atlas-content-modeler"
						)
				);
			}
		});
		reader.readAsText(uploadedFile);
	}

	/**
	 * Handle file change
	 * @param event
	 */
	async function onChangeHandler(event) {
		event.preventDefault();
		const uploadedFile = event.target.files[0];
		if (uploadedFile) {
			await validateFileUpload(uploadedFile);
		}
	}

	return (
		<>
			<input
				type="file"
				id="file"
				accept={allowedMimeTypes}
				ref={fileUploaderRef}
				className="hidden"
				onChange={onChangeHandler}
			/>

			<DarkButton
				data-testid="export-models-button"
				onClick={(event) => importClickHandler(event)}
			>
				{buttonTitle || __("Select File", "atlas-content-modeler")}
			</DarkButton>
		</>
	);
}
