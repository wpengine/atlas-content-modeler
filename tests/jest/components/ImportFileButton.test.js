import ImportFileButton from "../../../includes/settings/js/src/components/ImportFileButton";
import ShallowRenderer from "react-test-renderer/shallow";
import React from "react";

describe("ImportFileButton tests", () => {
	const fileData = {};
	const renderer = new ShallowRenderer();
	renderer.render(<ImportFileButton fileData={fileData} />);

	const tree = renderer.getRenderOutput();

	it("Renders a matching snapshot", () => {
		expect(tree).toMatchSnapshot();
	});
});
