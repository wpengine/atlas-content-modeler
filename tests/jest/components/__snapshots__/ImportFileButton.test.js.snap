// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportFileButton tests Renders a matching snapshot 1`] = `
<React.Fragment>
  <input
    className="hidden"
    id="file"
    onChange={[Function]}
    type="file"
  />
  <DarkButton
    data-testid="export-models-button"
    onClick={[Function]}
  >
    Select File
  </DarkButton>
</React.Fragment>
`;
